packages/bot-web-ui/src/stores/toolbox-store.js

```diff
@@ -0,0 +1,363 @@
+import { observable, action, reaction } from 'mobx';
+import { isMobile } from '@deriv/shared';
+import { localize } from '@deriv/translations';
+import { scrollWorkspace } from '@deriv/bot-skeleton';
+
+export default class ToolboxStore {
+    constructor(root_store) {
+        this.root_store = root_store;
+    }
+
+    @observable is_toolbox_open = true;
+    @observable is_search_loading = false;
+    @observable is_search_focus = false;
+    @observable sub_category_index = [];
+    @observable toolbox_dom = null;
+
+    @action.bound
+    onMount(toolbox_ref) {
+        this.toolbox_dom = Blockly.Xml.textToDom(toolbox_ref?.current);
+
+        this.workspace = Blockly.derivWorkspace;
+        this.adjustWorkspace();
+        this.disposeToolboxToggleReaction = reaction(
+            () => this.is_toolbox_open,
+            is_toolbox_open => {
+                if (is_toolbox_open) {
+                    this.adjustWorkspace();
+                }
+            }
+        );
+    }
+
+    @action.bound
+    onUnmount() {
+        if (typeof this.disposeToolboxToggleReaction === 'function') {
+            this.disposeToolboxToggleReaction();
+        }
+    }
+
+    @action.bound
+    adjustWorkspace() {
+        const { core } = this.root_store;
+
+        const toolbox_width = document.getElementById('gtm-toolbox')?.getBoundingClientRect().width || 0;
+        const block_canvas_rect = this.workspace.svgBlockCanvas_.getBoundingClientRect(); // eslint-disable-line
+
+        if (Math.round(block_canvas_rect.left) <= toolbox_width) {
+            const scroll_distance = isMobile() ? toolbox_width + 20 : toolbox_width + 36;
+            scrollWorkspace(this.workspace, scroll_distance, true, false);
+        }
+
+        // Emit event to GTM.
+        const { gtm } = core;
+        gtm.pushDataLayer({ event: 'dbot_toolbox_visible', value: true });
+    }
+
+    @action.bound
+    toggleDrawer() {
+        this.is_toolbox_open = !this.is_toolbox_open;
+    }
+
+    @action.bound
+    onToolboxItemClick(category) {
+        const { flyout } = this.root_store;
+        const category_id = category.getAttribute('id');
+        const dynamic = category.getAttribute('dynamic');
+        let flyout_content = Array.from(category.childNodes);
+
+        if (typeof dynamic === 'string') {
+            const fnToApply = this.workspace.getToolboxCategoryCallback(dynamic);
+            flyout_content = fnToApply(this.workspace);
```

Comment:
We seem to be using the same thing in a couple places, can we create a function for this?

Reply:
Good one. This is also done 👍